### ISSUE TITLE
Stable upstream branch?

### ISSUE THREAD

**Issue by lukego** [NixOS/nixpkgs#27385]
Hi! Just me again with another git workflow question :-).

I am interested in maintaining my own long-lived production branch of nixpkgs (#27312). I am looking for the most suitable upstream branch to pull changes from.

The most obvious branch is `master`. The downside is that this is the unstable branch, and I would actually value stability more than having the very latest version of each package.

The alternative that I see is the latest release branch like `release-17.03`. This seems to contain the software that I want to use. However, if I base my branch on `release-17.03` now then I am not sure how I will be able to upgrade to `17.09` and `18.03` ... in the future. Seems like each release branch is a dead-end upstream i.e. the next release branch will be cut from `master` and have a completely separate git history to the previous release branch.

Is that correct?

Assuming it is...

This would make it a challenge to merge new releases back into older ones. The cherry-picked changes would all conflict with the upstream versions. For example when I do `git checkout release-17.03 && git merge master` I see ~300 conflicts ([gist](https://gist.github.com/lukego/f66757556820bb314ad0d765bedf1c7b)) and would not be able to resolve those myself.

So... You know the old saying: The reasonable man adapts himself to the world; the unreasonable one persists in trying to adapt the world to himself.

The reasonable for me to do is to base my branch on the unstable `master` branch because that is the only long-lived branch available upstream. This way I can always merge with git. The catch is that I am running the unstable branch and presumably have more chance of breakage from a merge.

The unreasonable thing to do would be to suggest this git workflow for nixpkgs:

- There are two main upstream branches, `master` and `stable`.
- New work is based on `master` and merges there.
- Fixes to existing work are based on `stable` and merged both there and also to master.
- Release is made by merging master into stable.

This way a user can make a checkout of either the `stable` branch or the `master` branch and keep on pulling from it indefinitely. That is - the branch will never become a dead-end and require a rebase (which would not be an option for long-lived branches downstream.)

Sorry to confuse if I have misread the situation!

**Comment by vcunat** [comment]
Yes, stable branches are cut from master and they're dead-end.  The other way you suggest seems less flexible, e.g. I can't see how to well do changes that are specific to a stable branch.  For example, a CVE might be solved by update on master and by a patch on stable.

**Comment by lukego** [comment]
I actually suspect there is a straightforward way to use git-fu to construct a continuous `stable` branch from the release branches. Have to think about the details...

> I can't see how to well do changes that are specific to a stable branch. For example, a CVE might be solved by update on master and by a patch on stable.

I suppose that when you make the same "logical" change on two branches in different ways, then this will be a conflict to resolve when you merge (release.) So there will still be conflicts but maybe/hopefully 30 instead of 300?

**Comment by vcunat** [comment]
You may use a similar workflow yourself without changing the one of nixpkgs:

1. Start a branch on `git merge-base master release-17.03` and keep adding your changes there continually.
2. Maintain another branch as a merge of your branch with `release-17.03` (re-merged periodically).  This is the nixpkgs you want to evaluate.
3. Upgrade to 17.09: `git merge $(git merge-base master release-17.09)` to your branch.

I expect this would work reasonably if your changes don't (much) conflict with changes introduced in stable branches after they've been cut, i.e. when (2) is easy.  BTW, if your changes tend to conflict a lot with changes done on stable branches, you will probably have a hard time upgrading to next stable, whatever approach you choose.

**Comment by lukego** [comment]
How about this for an idea to construct a `stable` branch that always has exactly the content of the latest release branch (when sync'd) but includes the history of all past releases too.

Based on a trick: we want the _history_ of every release branch but we only want the _content_ of the latest one. So we want to merge (say) `release-16.09` and `release-17.03` and get a merge commit that has both as parents but has exactly the content of `release-17.03` (not even checking for conflicts.)

Git _almost_ supports this out of the box: see [stack overflow question](https://stackoverflow.com/questions/4911794/git-command-for-making-one-branch-like-another/4912267#4912267). Basically when `release-17.09` is created we will want to run `git merge -s theirs release-17.09` to overwrite the contents of our branch with the new release while still keeping all the histories. The only catch is that `--theirs` doesn't exist and has to be emulated with `--ours` on a temp branch.

Here is a proof of concept shell script to bootstrap the initial copy of the stable branch (done only once):

```
# Create branch 'stable' with the contents of the latest release
# branch and the history of all the previous release branches.
git checkout -B stable release-17.03
for release in 16.09 16.03 15.09 14.12 14.04 13.10; do
    git merge -m "Merge release $release into stable" \
        -s ours origin/release-${release}
done
```

This creates a branch that merges all existing releases:

```
$ git log --oneline -5
ab46978 Merge release 13.10 into stable
1187c16 Merge release 14.04 into stable
6e246f4 Merge release 14.12 into stable
81c2840 Merge release 15.09 into stable
7715100 Merge release 16.09 into stable
```

and includes all the commits from those branches:

```
$ git log HEAD..origin/release-13.10
$
```

and is identical to the latest release branch:

```
$ git diff stable origin/release-17.03
$
```

So - now we have a stable branch with the past available. If you have a branch based on an old release, e.g. 13.10, you should be able to upgrade that by pulling from this branch - and only resolve conflicts directly related to changes that you made downstream.

Going forward we could maintain this branch by (a) merging from the latest release branch regularly and (b) each time a new release is made we use the stackoverflow trick above to adopt its content while preserving our history.

Cool, huh?

I pushed the branch to [lukego/stable](https://github.com/lukego/nixpkgs/tree/stable) in case anybody wants to take a peek and help decide whether this is something that should be maintained for realz.

**Comment by lukego** [comment]
@vcunat Could be that I am missing something, I haven't used `merge-base`, but in that example won't the person maintaining the downstream branch have to resolve lots of conflicts between changes that have been made separately on each upstream branch? That is the problem that I want to avoid.

**Comment by vcunat** [comment]
@lukego: `git merge-base` will in our case give the forking point, i.e. the last common ancestor of both stable and master.

**Comment by vcunat** [comment]
I think your suggestion is likely to create some fewer conflicts, specifically on files where exactly the same changes have been picked from master to stable.  Those could be auto-solved by `git-rerere` in the approach I suggested, but it seems better to get rid of them.

I believe you can easily use the approach even without having an official branch like this.  Still, there could be one, as there won't be much work due to periodical re-merges due to multiple stable branches getting commits "in parallel" sometimes.

**Comment by lukego** [comment]
Thanks for the discussion! Closing now because it seems like it is actually safe to start a long-lived branch from any of the release branches. Some special action will be needed to upgrade to the following release but it will only take a few carefully chosen git commands.

I agree the 'stable' branch is not fundamentally needed, just a potential convenience that could be adopted at any time.
